src/Discord/Internal/Types/ApplicationCommands.hs

```diff
@@ -0,0 +1,382 @@
+  toJSON = toJSON . fromEnum
+
+instance FromJSON ApplicationCommandType where
+  parseJSON = withScientific "ApplicationCommandType" (return . toEnum . round)
+
+-- | Data type to be used when creating application commands. The specification
+-- is below.
+--
+-- If a command of the same type and and name is sent to the server, it will
+-- overwrite any command that already exists in the same scope (guild vs
+-- global).
+--
+-- The description has to be empty for non-slash command application
+-- commands, as do the options. The options need to be `Nothing` for non-slash
+-- commands, too. If one of the options is a subcommand or subcommand group,
+-- the base command will no longer be usable.
+--
+-- A subcommand group can have subcommands within it. This is the maximum amount
+-- of command nesting permitted.
+--
+-- https://discord.com/developers/docs/interactions/application-commands#create-global-application-command
+data CreateApplicationCommand = CreateApplicationCommand
+  { -- | The application command name (1-32 chars).
+    createApplicationCommandName :: String,
+    -- | The application command description (1-100 chars). Has to be empty for
+    -- non-slash commands.
+    createApplicationCommandDescription :: String,
+    -- | What options the application (max length 25). Has to be `Nothing` for
+    -- non-slash  commands.
+    createApplicationCommandOptions :: Maybe [ApplicationCommandOption],
+    -- | Whether the command is enabled by default when the application is added
+    -- to a guild. Defaults to true if not present
+    createApplicationCommandDefaultPermission :: Maybe Bool,
+    -- | What the type of the command is. If `Nothing`, defaults to slash
+    -- commands.
+    createApplicationCommandType :: Maybe ApplicationCommandType
+  }
+  deriving (Show, Eq, Read)
+
+instance Default CreateApplicationCommand where
+  def = CreateApplicationCommand "createappcom" "createappcom desc" Nothing Nothing Nothing
+
+instance ToJSON CreateApplicationCommand where
+  toJSON CreateApplicationCommand {..} =
+    object
+      [ (name, value)
+        | (name, Just value) <-
+            [ ("name", toMaybeJSON createApplicationCommandName),
+              ("description", toMaybeJSON createApplicationCommandDescription),
+              ("options", toJSON <$> createApplicationCommandOptions),
+              ("default_permission", toJSON <$> createApplicationCommandDefaultPermission),
+              ("type", toJSON <$> createApplicationCommandType)
+            ]
+      ]
+
+-- | Data type to be used when editing application commands. The specification
+-- is below. See `CreateApplicationCommand` for an explanation for the
+-- parameters.
+--
+-- https://discord.com/developers/docs/interactions/application-commands#edit-global-application-command
+data EditApplicationCommand = EditApplicationCommand
+  { editApplicationCommandName :: Maybe String,
+    editApplicationCommandDescription :: Maybe String,
```

Comment:
Use Text instead of String. (I prefer to use text for types as much as possible)